crates/ibc/src/actions.rs

```diff
@@ -4,77 +4,244 @@ use std::cell::RefCell;
-    D: DB + for<'iter> DBIter<'iter>,
-    H: StorageHasher,
+    S: State,
 {
-    type D = D;
-    type H = H;
+    type PrefixIter<'iter> = <S as StorageRead>::PrefixIter<'iter> where Self: 'iter;
 
-    fn write_log(&self) -> &WriteLog {
-        self.wl_storage.write_log()
+    fn read_bytes(
+        &self,
+        key: &namada_storage::Key,
+    ) -> StorageResult<Option<Vec<u8>>> {
+        self.state.read_bytes(key)
+    }
+
+    fn has_key(&self, key: &namada_storage::Key) -> StorageResult<bool> {
+        self.state.has_key(key)
     }
 
-    fn write_log_mut(&mut self) -> &mut WriteLog {
-        self.wl_storage.write_log_mut()
+    fn iter_prefix<'iter>(
+        &'iter self,
+        prefix: &namada_storage::Key,
+    ) -> StorageResult<Self::PrefixIter<'iter>> {
+        self.state.iter_prefix(prefix)
     }
 
-    fn storage(&self) -> &State<D, H> {
-        self.wl_storage.storage()
+    fn iter_next<'iter>(
+        &'iter self,
+        iter: &mut Self::PrefixIter<'iter>,
+    ) -> StorageResult<Option<(String, Vec<u8>)>> {
+        self.state.iter_next(iter)
     }
 
-    fn split_borrow(&mut self) -> (&mut WriteLog, &State<D, H>) {
-        self.wl_storage.split_borrow()
+    fn get_chain_id(&self) -> StorageResult<String> {
+        self.state.get_chain_id()
     }
 
-    fn write_tx_hash(&mut self, hash: Hash) -> write_log::Result<()> {
-        self.wl_storage.write_tx_hash(hash)
+    fn get_block_height(&self) -> StorageResult<namada_storage::BlockHeight> {
+        self.state.get_block_height()
+    }
+
+    fn get_block_header(
+        &self,
+        height: namada_storage::BlockHeight,
+    ) -> StorageResult<Option<namada_storage::Header>> {
+        StorageRead::get_block_header(self.state, height)
+    }
+
+    fn get_block_hash(&self) -> StorageResult<namada_storage::BlockHash> {
+        self.state.get_block_hash()
+    }
+
+    fn get_block_epoch(&self) -> StorageResult<namada_storage::Epoch> {
+        self.state.get_block_epoch()
+    }
+
+    fn get_pred_epochs(&self) -> StorageResult<Epochs> {
+        self.state.get_pred_epochs()
+    }
+
+    fn get_tx_index(&self) -> StorageResult<namada_storage::TxIndex> {
+        self.state.get_tx_index()
+    }
+
+    fn get_native_token(&self) -> StorageResult<Address> {
+        self.state.get_native_token()
     }
 }
-namada_state::impl_storage_traits!(IbcProtocolContext<'_, D, H>);
 
-impl<D, H> IbcStorageContext for IbcProtocolContext<'_, D, H>
+impl<S> StorageWrite for IbcProtocolContext<'_, S>
 where
-    D: DB + for<'iter> DBIter<'iter> + 'static,
-    H: StorageHasher + 'static,
+    S: State,
+{
+    fn write_bytes(
+        &mut self,
+        key: &namada_storage::Key,
+        val: impl AsRef<[u8]>,
+    ) -> StorageResult<()> {
+        self.state.write_bytes(key, val)
+    }
+
+    fn delete(&mut self, key: &namada_storage::Key) -> StorageResult<()> {
+        self.state.delete(key)
+    }
+}
+
+// impl<S> StateRead for IbcProtocolContext<'_, S>
```

Comment:
I think we can remove this commented block if not needed

Reply:
done in c8f43714b0381ea0ac091a09dade646d69109135